api/python/quilt3/data_transfer.py

```diff
@@ -897,7 +925,124 @@ def calculate_sha256(src_list: List[PhysicalKey], sizes: List[int]):
     return _calculate_sha256_internal(src_list, sizes, [None] * len(src_list))
 
 
-def _calculate_hash_get_s3_chunks(ctx, src, size):
+def with_lock(f):
+    lock = threading.Lock()
+
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        with lock:
+            return f(*args, **kwargs)
+    return wrapper
+
+
+@retry(stop=stop_after_attempt(MAX_FIX_HASH_RETRIES),
+       wait=wait_exponential(multiplier=1, min=1, max=10),
+       retry=retry_if_result(lambda results: any(r is None or isinstance(r, Exception) for r in results)),
+       retry_error_callback=lambda retry_state: retry_state.outcome.result(),
+       )
+def _calculate_sha256_internal(src_list, sizes, results):
+    total_size = sum(
+        size
+        for size, result in zip(sizes, results)
+        if result is None or isinstance(result, Exception)
+    )
+    stopped = False
+
+    with tqdm(desc="Hashing", total=total_size, unit='B', unit_scale=True, disable=DISABLE_TQDM) as progress, \
+         ThreadPoolExecutor(MAX_CONCURRENCY) as executor:
+
+        find_correct_client = with_lock(S3ClientProvider().find_correct_client)
+        progress_update = with_lock(progress.update)
+
+        def _process_url_part(src: PhysicalKey, offset: int, length: int):
+            hash_obj = hashlib.sha256()
+
+            if src.is_local():
+                bytes_remaining = length
+                with open(src.path, 'rb') as fd:
+                    fd.seek(offset)
+                    while bytes_remaining > 0:
+                        chunk = fd.read(min(s3_transfer_config.io_chunksize, bytes_remaining))
+                        if not chunk:
+                            # Should not happen, but let's not get stuck in an infinite loop.
+                            break
+                        hash_obj.update(chunk)
+                        progress_update(len(chunk))
+                        bytes_remaining -= len(chunk)
+            else:
+                end = offset + length - 1
+                params = dict(
+                    Bucket=src.bucket,
+                    Key=src.path,
+                    Range=f'bytes={offset}-{end}',
+                )
+                if src.version_id is not None:
+                    params.update(VersionId=src.version_id)
+
+                s3_client = find_correct_client(S3Api.GET_OBJECT, src.bucket, params)
+
+                try:
+                    body = s3_client.get_object(**params)['Body']
```

Comment:
@dimaryaz we can use MPU hack here to compute checksums in the cloud w/o streaming the bytes, tho we need some scratch location to host that MPU

Reply:
we could configure service bucket during `quilt3 login`
but in that case it might be a better idea to just use backend
I wonder if we even move (at least part of) pkgpush logic to quilt3 🤔🤔🤔